substrate/frame/staking/src/lib.rs

```diff
@@ -1229,31 +1291,97 @@ impl<T: Config> EraInfo<T> {
 	}
 
 	/// Store exposure for elected validators at start of an era.
-	pub fn set_exposure(
+	///
+	/// If the exposure does not exist yet for the tuple (era, validator), it sets it. Otherwise,
+	/// it updates the existing record by ensuring *intermediate* exposure pages are filled up with
+	/// `T::MaxExposurePageSize` number of backers per page and the remaining exposures are added
+	/// to new exposure pages.
+	pub fn upsert_exposure(
 		era: EraIndex,
 		validator: &T::AccountId,
-		exposure: Exposure<T::AccountId, BalanceOf<T>>,
+		mut exposure: Exposure<T::AccountId, BalanceOf<T>>,
 	) {
 		let page_size = T::MaxExposurePageSize::get().defensive_max(1);
 
-		let nominator_count = exposure.others.len();
-		// expected page count is the number of nominators divided by the page size, rounded up.
-		let expected_page_count = nominator_count
-			.defensive_saturating_add((page_size as usize).defensive_saturating_sub(1))
-			.saturating_div(page_size as usize);
-
-		let (exposure_metadata, exposure_pages) = exposure.into_pages(page_size);
-		defensive_assert!(exposure_pages.len() == expected_page_count, "unexpected page count");
+		if let Some(stored_overview) = ErasStakersOverview::<T>::get(era, &validator) {
```

Comment:
@Ank4n @kianenigma to review this one last time. I think it can be written simpler. 

Does it have good tests?